### PR TITLE
Unconditionally increment version counters for in-place ops

### LINKED ISSUE #14571: Inputs with output_differentiability=False don't bump version counter
Example (from autogenerated code):
```
 - name: no_grad_embedding_renorm_(Tensor self, Tensor indices, double max_norm, double norm_type)
   output_differentiability: [False, False, False, False]
```
```
Tensor & VariableType::no_grad_embedding_renorm_(Tensor & self, const Tensor & indices, double ma
x_norm, double norm_type) const {
  profiler::RecordFunction profiler("no_grad_embedding_renorm_", Function::peek_at_next_sequence_
nr());
  auto& self_ = unpack(self, "self", 0);
  auto& indices_ = unpack(indices, "indices", 1);
  torch::jit::Node* node = nullptr;
  std::shared_ptr<jit::tracer::TracingState> tracer_state;
  if (jit::tracer::isTracing()) {
    tracer_state = jit::tracer::getTracingState();
    at::Symbol op_name;
    if (tracer_state->force_outplace) {
      op_name = jit::Symbol::fromQualString("aten::no_grad_embedding_renorm");
    } else {
      op_name = jit::Symbol::fromQualString("aten::no_grad_embedding_renorm_");
    }
    node = tracer_state->graph->create(op_name, /*num_outputs=*/0);
    jit::tracer::recordSourceLocation(node);
    jit::tracer::addInputs(node, "self", self);
    jit::tracer::addInputs(node, "indices", indices);
    jit::tracer::addInputs(node, "max_norm", max_norm);
    jit::tracer::addInputs(node, "norm_type", norm_type);
    tracer_state->graph->appendNode(node);
    jit::tracer::ensureUniqueIfOutOfPlaced("no_grad_embedding_renorm_", self);
    jit::tracer::setTracingState(nullptr);
  }
  baseType->no_grad_embedding_renorm_(self_, indices_, max_norm, norm_type);
  if (tracer_state) {
    jit::tracer::setTracingState(std::move(tracer_state));
    jit::tracer::addOutput(node, self);
  }
  return self;
}
```

We don't bump the version counter of `self` even though it is modified in place. This sounds dangerous; we could have used `self` in another computation where it is being saved for some Function's backward pass.

cc @colesbury @yf225

### PR DESCRIPTION
**Pull request by zou3519**
Fixes #14571

Before we were not incrementing version counters for in-place operations
where the input was marked as not requiring derivative. This PR changes
it so that we always increment the version counters for inputs that are
modified by in-place operations.

Before: https://gist.github.com/zou3519/a9d24afc6ea4aaf705f9bafc8015aa0a
After: https://gist.github.com/zou3519/24021fd13f31f090ecd319b3f22c7a28

Test Plan:
- Read generated code.

### CODE REVIEW

**Comment by soumith** [comment]
add a test?

**Comment by zou3519** [comment]
Err this isn't the correct solution; because the only place where this is a problem is no_grad_embedding_renorm_. Closing this in favor of fixing that one function.

**Comment by apaszke** [comment]
Why isn't that the correct solution? The fact that this problem was surfaced by a single function doesn't mean it won't appear in the future...

**Comment by SsnL** [comment]
I agree with @apaszke. We should fix the codegen.

**Comment by zou3519** [comment]
If we only use output_differentiability=False on tensors that are never exposed to the user, then we don't have a problem unless a pytorch dev modifies a tensor in place. If we say that output_differentiability should be user-facing, then I agree we should fix the codegen.

At any rate the pr in its current iteration failed a few tests so there is something else I'm not understanding

**Comment by apaszke** [comment]
Well if the real meaning is "this tensor is not exposed to the user" then we should at least change the name, because its current way only encourages mistakes...

**Comment by SsnL** [comment]
Hmm, so any input that is modified inplace and doesn't require derivative will have this bug? Then shouldn't this also include batch_mean/var of batch norm? Then it doesn't seem that uncommon. IMO, we should fix the code & the bugs it uncovers.

**Comment by zou3519** [comment]
Any input that is modified in place, marked as not requiring derivative in the codegen (I don't fully understand when this happens. The most common case is when output_differentiability=False is set, but we also have some functions that we exclude from requiring derivatives), and exposed to the user will have this problem.

Are batch_mean/var of batch norm marked as not requiring derivative in the codegen?
